packages/at_secondary_server/test/update_verb_test.dart

```diff
@@ -914,6 +921,187 @@ void main() {
       }
     });
 
+    test('Concurrent updates to the same IMMUTABLE record', () async {
+      int concurrency = 10;
+      // Kicks off 10 update verbs concurrently to ensure that the concurrent
+      // update mutex protection is working correctly.
+      UpdateVerbHandler updateHandler = UpdateVerbHandler(
+          secondaryKeyStore, statsNotificationService, notificationManager);
+
+      final atKey =
+          '${DateTime.now().millisecondsSinceEpoch}.concurrent.tests@alice';
+      List<InboundConnection> connections = [];
+      List<Future<void>> futures = [];
+      for (int i = 0; i < concurrency; i++) {
+        var c = DummyInboundConnection();
+        c.metaData.isAuthenticated = true;
+        connections.add(c);
+        futures.add(updateHandler.process(
+            'update:immutable:true:$atKey original data', c));
+      }
+      int successes = 0;
+      int failures = 0;
+      for (int i = 0; i < concurrency; i++) {
+        if (i > 0) {
+          expect(updateHandler.updateMutexes.length, 1);
+        }
+        await futures[i]
+            .then((value) => successes++)
+            .catchError((e, st) => failures++);
+      }
+
+      // Because this is an immutable record we expect only 1 success
+      expect(successes, 1);
+      expect(failures, concurrency - 1);
+      expect(updateHandler.updateMutexes.length, 0);
+
+      AtData? data = await secondaryKeyStore.get(atKey);
+      expect(data, isNotNull);
+      expect(data!.metaData, isNotNull);
+      // Because this is an immutable record we expect version to be `0`
+      expect(data.metaData!.version, 0);
+    });
+
+    test('Concurrent updates to the same MUTABLE record', () async {
```

Comment:
same check as the previous test except for normal (mutable) records. This time we expect all updates to succeed, and we expect the `AtMetaData.version` to be 9 after the last one (because version starts at 0)